File: general-interview-prep/javascript/find132Pattern.js
/**
 * @param {number[]} nums
 * @return {boolean}
 */
var find132pattern = function(nums) {
    // use a stack

    // then pop off stack and
    // check base case

    let max = -Infinity;
    let stack = [];

    for(let i = nums.length - 1; i >= 0; i--) {
        while(stack.length && stack[stack.length - 1] < nums[i]) {
            max = Math.max(max, stack[stack.length - 1]);
            stack.pop();
        }

        if(max !== -Infinity && max > nums[i] && stack.length) return true;

        if(!stack.length || (nums[i] < stack[stack.length - 1])) {
            stack.push(nums[i]);
        }
    }

    return false;
};

File: general-interview-prep/golang/find132Pattern.go
func find132pattern(nums []int) bool {
    max := math.Inf(-1)
    stack := []int{}

    for i := len(nums) - 1; i >= 0; i-- {
        for len(stack) != 0 && nums[i] > stack[len(stack) - 1] {
            max = math.Max(max, float64(stack[len(stack) - 1]))
            stack = stack[:len(stack) - 1]
        }

        if max != math.Inf(-1) && nums[i] < int(max) && len(stack) != 0 {
            return true
        }

        if len(stack) == 0 || nums[i] < stack[len(stack) - 1] {
            stack = append(stack, nums[i])
        }
    }

    return false
}
